ivy/functional/backends/paddle/device.py

```diff
@@ -8,71 +8,118 @@
     /,
     *,
     stream: Optional[int] = None,
     out: Optional[paddle.Tensor] = None,
 ) -> paddle.Tensor:
-    if device is None:
-        return x
-    # TODO: check memory leak because ret is copying the original tensor
-    ret = paddle.to_tensor(
-        x, place=as_native_dev(device), stop_gradient=x.stop_gradient
-    )
-    return ret
+    if device is not None:
+        native_dev = as_native_dev(device)
+        if not dev(x, as_native=True)._equals(native_dev):  # type: ignore
+            # TODO: check memory leak because ret is copying the original tensor
+            return paddle.to_tensor(x, place=native_dev, stop_gradient=x.stop_gradient)
+    return x
 
 
-def as_ivy_dev(device: Place, /):
+def as_ivy_dev(
+    device: Union[str, _paddle_dev_types],
+    /,
+) -> ivy.Device:
     if isinstance(device, str):
         return ivy.Device(device)
 
-    if device.is_cpu_place():
-        dev_type = "cpu"
-        dev_idx = 0
-    elif device.is_gpu_place():
-        dev_type = "gpu"
-        dev_idx = device.gpu_device_id()
-
-    if dev_type == "cpu":
-        return ivy.Device(dev_type)
-
-    return ivy.Device(dev_type + (":" + (str(dev_idx) if dev_idx is not None else "0")))
+    if is_native_dev(device):
+        if not isinstance(device, Place):
+            native_dev = Place()
+            native_dev.set_place(device)
+        else:
+            native_dev = device
+        if native_dev.is_cpu_place():
+            return ivy.Device("cpu")
+        elif native_dev.is_gpu_place():
+            dev_type = "gpu"
+            dev_idx = device.gpu_device_id()
+            return ivy.Device(dev_type + ":" + str(dev_idx))
+        elif native_dev.is_gpu_pinned_place():
+            return ivy.Device("gpu:0")  # simiplification
+        else:
+            raise ivy.utils.exceptions.IvyError(
+                f"Cannot convert {device} to an ivyDevice. ",
+                "Currently only cpu, tpu and gpu are supported.",
+            )
+
+    raise ivy.utils.exceptions.IvyError(
+        f"Cannot convert {device} to an ivy device. Expected a "
+        f"paddle.fluid.libpaddle.Place or str, got {type(device)}"
+    )
 
 
 def as_native_dev(
-    device: Optional[Union[ivy.Device, Place]] = None,
+    device: Union[ivy.Device, str, _paddle_dev_types],
     /,
-) -> Optional[Place]:
-    if not isinstance(device, str):
+) -> _paddle_dev_types:
+    if is_native_dev(device):
         return device
-    elif "cpu" in device:
-        return paddle.fluid.libpaddle.CPUPlace()
-    elif "gpu" in device:
-        return paddle.fluid.libpaddle.CUDAPlace()
+    if isinstance(device, str):
+        if "tpu" in device:
+            raise ivy.utils.exceptions.IvyBackendException(
+                "TPU not supported in paddle backend. Consider using ",
+                "TensorFlow or JaX.",
+            )
+        else:
+            # Internal function handles more edge cases and devices
+            return _get_paddle_place(device)
```

Comment:
It's preferable that we copy this internal function and not import it to avoid breaking changes later as this is not supposed to be directly used, I assume that it could be much simpler after we restrict it to cpu/gpu only as we don't support other devices anyway